experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/objectFactory.ts

```diff
@@ -52,6 +44,63 @@ export function addFactory<TSchema extends ObjectNodeSchema>(
 	}) as FactoryTreeSchema<TSchema>;
 }
 
+/**
+ * Given a content tree that is to be inserted into the shared tree, replace all subtrees that were created by factories
+ * (via {@link SharedTreeObjectFactory.create}) with the content that was passed to those factories.
+ * @remarks
+ * This functions works recursively.
+ * Factory-created objects that are nested inside of the content passed to other factory-created objects, and so on, will be in-lined.
+ * This function also adds the hidden {@link typeNameSymbol} of each object schema to the output.
+ * @example
+ * ```ts
+ * const x = foo.create({
+ *   a: 3, b: bar.create({
+ *     c: [baz.create({ d: 5 })]
+ *   })
+ * });
+ * const y = extractFactoryContent(y);
+ * y === {
+ *   [typeNameSymbol]: "foo", a: 3, b: {
+ *     [typeNameSymbol]: "bar", c: [{ [typeNameSymbol]: "baz", d: 5 }]
+ *  }
+ * }
+ * ```
+ */
+export function extractFactoryContent<T extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	content: T,
+): T {
+	const copy: Record<string, unknown> = {};
+	if (Array.isArray(content)) {
+		// `content` is an array
+		return content.map(extractFactoryContent) as T;
+	} else if (content !== null && typeof content === "object") {
+		const factoryContent = (content as Partial<HasFactoryContent<object>>)[
+			factoryContentSymbol
+		];
+		if (factoryContent !== undefined) {
+			// `content` is a factory-created object
+			const typeName =
+				(factoryContent as { [typeNameSymbol]?: string })[typeNameSymbol] ??
+				fail("Expected schema type name to be set on factory object content");
+
+			Object.defineProperty(copy, typeNameSymbol, { value: typeName });
+			for (const [p, v] of Object.entries(factoryContent)) {
+				copy[p] = extractFactoryContent(v);
+			}
+		} else {
+			// `content` is a POJO object (but may have factory-created objects within it)
```

Comment:
Nit: there are a couple of places where "POJO" is used, but I wouldn't have necessarily known what that meant without chatting. Might be worth expanding the abbreviations for clarity.